finql-sqlite/src/transaction_handler.rs

```diff
@@ -136,92 +134,134 @@ impl TransactionHandler for SqliteDB {
+                sql.push_str(" AND asset_id = ?2");
+            }
+            else {
+                sql.push_str(" AND asset_id IS NULL");
+            }
+            sql.push_str(" AND cash_amount = ?3 AND cash_currency_id = ?4 AND cash_date = ?5");
+
+            let id = conn.query_row(
+                &sql,
+                params![
+                    &transaction.trans_type,
+                    &transaction.asset,
+                    &transaction.cash_amount,
+                    &transaction.cash_currency.id.expect("cash currency asset id must be in db"),
+                    &transaction.cash_date,
                 ],
-                |row| row.get(0) )?)
+                |row| {
+                    let id:usize = row.get(0)?;
+                    Ok(id)
+                })?;
+
+            Ok(id)
         }).await.map_err(|e| DataError::DataAccessFailure(e.to_string()))?
         .map_err(|e| DataError::DataAccessFailure(e.to_string()))
     }
 
     async fn get_transaction_by_id(&self, id: usize) -> Result<Transaction, DataError> {
         Ok(self.conn.interact(move |conn| -> Result<RawTransaction, SQLiteError> {
             Ok(conn.query_row(
-                "SELECT trans_type, asset_id, 
-                cash_amount, cash_currency, cash_date, related_trans, position, note 
-                FROM transactions
-                WHERE id=?1",
+                "SELECT
+                t.id,
+                t.trans_type,
+                t.asset_id,
+                t.cash_amount,
+                c.id AS cash_currency_id,
+                c.iso_code AS cash_iso_code,
+                c.rounding_digits AS cash_rounding_digits,
+                t.cash_date,
+                t.related_trans,
+                t.position,
+                t.note
+                FROM transactions t
+                JOIN currencies c ON c.id = t.cash_currency_id
+                WHERE t.id = ?1",
                 params![&id],
-                |row| { Ok(RawTransaction {
-                    id: Some(id),
-                    trans_type: row.get(0)?,
-                    asset: row.get(1)?,
-                    cash_amount: row.get(2)?,
-                    cash_currency: row.get(3)?,
-                    cash_date: row.get(4)?,
-                    related_trans: row.get(5)?,
-                    position: row.get(6)?,
-                    note: row.get(7)?
-                })
-            })?)
+                |row| {
+                    let cic:String = row.get(5)?;
+
+                    Ok(RawTransaction {
+                        id: row.get(0)?,
+                        trans_type: row.get(1)?,
+                        asset: row.get(2)?,
+                        cash_amount: row.get(3)?,
+                        cash_currency: Currency::new(
+                            row.get(4)?,
+                            CurrencyISOCode::from_str(&cic).expect("expected good code from the database"),
+                            row.get(6)?,
```

Comment:
Please no expect or unwrap anywhere in the core library! Library code that is used in other programs should never panic. It's okay to unwrap if the value has been checked previously to be okay, but not otherwise. The database entry could be corrupted, therefore I would return an error instead of using `expect`. 
In tests and examples `unwrap` or `expect` is no problem.